text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,203 @@
+region and keep the old, and possible stale, bucket keys. The point is that bucket keys don't need
+to be 100% accurate, a close match can enhance concurrency and load balance a lot. For fast lookup,
+a map from region id to bucket keys can also be maintained.
+
+After a region becomes larger, it needs to be warmed up before serving requests. There are two
+common cases:
+
+- In the very beginning, writes will only happen in a single region, the problem can be solved by
+    pre-split + scatter.
+- Because followers don’t serve read, so their data is usually not in cache. After transferring
+    the leader, serving a range scan can trigger at most 10GiB IO. To warm it up, we can extend
+    the existing pre transfer leader mechanism. Leader will send its bucket statics to follower
+    in pre transfer leader, follower can do a pre scan for hot buckets before accepting
+    transferring leader.
+
+### Flow report
+
+Buckets statistics are reported in a standalone stream. They will be reported more frequently
+than original region heartbeats, which is 60s. The new stream is expected to be reported every
+10 seconds when have changes.
+
+PD will maintain top N hot spot buckets and split all of those buckets that exceed P of
+all throughput. N and P should be configurable, I suggest to set it to 300 and 1% respectively
+first. The algorithm is just one possible solution and subject to change for evaluation.
+
+Because the boundaries of buckets can changed by writes, so PD may need to merge the history
+of different versions of buckets. There are two possible solutions. For example, buckets
+{ [a, c), [c, e) } is replaced by { [a, b), [b, d), [d, e) },
+1. To simulate the old split behavior, history of [a, c) should be inherited by [b, d) and [c, e)
+    should be inherited by [d, e). The history should be inherited by the rightest overlapping
+    buckets.
+2. Or the history of [a, c) can be shared by both [a, b) and [b, d), [d, e) can be shared by
+    both [b, d) and [d, e). The history should be inherited by all overlapping buckets and
+    optionally multiply a factor like 0.8 for errors.
+
+### Replication
+
+A large region can take minutes to be replicated, PD should make cost estimation on actual size
+instead of fixed value. To avoid logs being truncated during sending a snapshot, we should set
+the log gc limit based on the actual size too. Raft engine should be GA before dynamic size
+region, so we don't need to worry about the extra write amplifications.
+
+To avoid ingesting a large snapshot cause large compaction, we need to split the snapshot into
+several smaller files. Each snapshot files should not be larger than 128MiB.
+
+There are ways to optimize the replications, sending a sub-LSM tree for example. But these designs
+may be complicated. And we rely on further designs like separating LSM tree to optimize the cost
+for all.
+
+### Compatibility
```

Comment:
If we implement the feature in the way that buckets are just optional and optimization for big region size, then by its nature it's backward compatible. 
And there's nothing prevent user from using small region,  so this should be well supported in the long term as well.  For small to medium size cluster, it's likely user will keep the small region size.